R/ds.studentTestDP.R

```diff
@@ -0,0 +1,84 @@
+#'
+#' @return \code{ds.studentTestDP} returns a differentially private student-test
+#' @export
+
+
+ds.studentTestDP <- function(datasources, x, y, epsilon, x_min, x_max, y_min, y_max, type="combine") {
+    if (is.null(datasources)) {
+        datasources <- DSI::datashield.connections_find()
+    }
+
+    split_epsilon <- epsilon/2
+
+    cally_x <- paste0("calculationsForStudentTestDP(", x, ", ", split_epsilon, ", ", x_min, ", ", x_max, ")")
+    cally_y <- paste0("calculationsForStudentTestDP(", y, ", ", split_epsilon, ", ", y_min, ", ", y_max, ")")
+    res_x <- DSI::datashield.aggregate(datasources, as.symbol(cally_x))
+    res_y <- DSI::datashield.aggregate(datasources, as.symbol(cally_y))
+    
+    Nstudies <- length(datasources)
+    
+    studentTest.combine <- computeStudentTestCombine(Nstudies, res_x, res_y)
+    studentTest.split <- computeStudentTestSplit(Nstudies, res_x, res_y)
+
+    if (type == "both") {
+        return(list(StudentTest.by.Study=studentTest.split,Global.StudentTest=studentTest.combine,Nstudies=Nstudies))
+    } else if (type == "combine") {
+        return(list(Global.StudentTest=studentTest.combine,Nstudies=Nstudies))
+    } else if (type == "split") {
+        return(list(StudentTest.by.Study=studentTest.split,Nstudies=Nstudies))
+    } else {
+        stop("type must be either 'both', 'combine' or 'split'")
+    }
+}
+
+computeGlobalStats <- function(Nstudies, input) {
+    GlobalSum <- 0
+    GlobalSumSquares <- 0
+    GlobalNvalid <- 0  
+    for (i in 1:Nstudies){
+        GlobalSum <- GlobalSum +  input[[i]][[1]]
+        GlobalSumSquares <- GlobalSumSquares +  input[[i]][[2]]
+        GlobalNvalid <- GlobalNvalid +  input[[i]][[3]]
+    }
+
+    GlobalVar <- GlobalSumSquares/(GlobalNvalid-1) - (GlobalSum^2)/(GlobalNvalid*(GlobalNvalid-1))
+    GlobalMean <- GlobalSum/GlobalNvalid
+    GlobalNvalid <- GlobalNvalid
+    return (list(GlobalVar=GlobalVar,GlobalMean=GlobalMean,GlobalNvalid=GlobalNvalid))
+}
+
+computeStudentTestCombine <- function(Nstudies, input_x, input_y) {
+    x_stats <- computeGlobalStats(Nstudies, input_x) 
+    y_stats <- computeGlobalStats(Nstudies, input_y)
+    StudentTest <- (x_stats$GlobalMean - y_stats$GlobalMean)/sqrt(x_stats$GlobalVar / x_stats$GlobalNvalid + y_stats$GlobalVar / y_stats$GlobalNvalid)
+    return (StudentTest)
+}
+
+computeLocalStats <- function(input_col){
+    Nvalid <- as.numeric(input_col[[3]])
+    Mean <- input_col[[1]] / input_col[[3]]
+    Var <- input_col[[2]]/(Nvalid-1) - (input_col[[1]]^2)/(Nvalid*(Nvalid-1))
+    return (list(Nvalid=Nvalid,Mean=Mean,Var=Var))
+}
+
+computeStudentTestSplit <- function(Nstudies, input_x, input_y) {
+    StudentTest <- c()
+    for (i in 1:Nstudies){
+        x_stats <- computeLocalStats(input_x[[i]])
+        y_stats <- computeLocalStats(input_y[[i]])
+        StudentTest[i] <- (x_stats$Mean - y_stats$Mean)/sqrt(x_stats$Var / x_stats$Nvalid + y_stats$Var / y_stats$Nvalid) 
```

Comment:
This will be incorrect when `x` and `y` have invalid samples at different indexes. I think you should first start by "removing" index `i` from both inputs whenever `x[i]` OR `y[i]` is invalid. This have to be done on the server side, so it means that `calculationsForStudentTestDP` should be rewritten to operate both on `x` and `y`. The good news it that you will spare a bit of budget since you will have a single DP count instead of 2!

Reply:
Not sure this is true, in this case x and y are 2 groups of different individuals (not 2 variables on the same set of individuals as it was the case for the covariance)

Reply:
Hmmm you're damn right, I was too influenced by Cov! This is good for me then :)